Fix JTC state tolerance and goal_time tolerance check bug

mechwiz opened the pull request:
Hey all,

I noticed a bug in the JTC where if a joint trajectory tolerance was specified (usually done using the `constraints.<joint_name>.trajectory` param in the ros_controllers.yaml file) for a given joint, the trajectory point being processed still gets set to the relevant command interfaces **before** getting tolerance checked. If in fact the tolerance is violated (which is determined afterwards), the offending joint command still gets sent out which we don't want. The fix is simply to tolerance check first and then set the command interfaces if the tolerance is met. 

I came across this bug specifically for the first point in the trajectory (but I think this could happen to subsequent points depending on how the trajectory was made). Usually the first point is checked to be within the "allowed_start_tolerance" defined in MoveIt (so the tolerance is usually met when it gets to the JTC) but in this scenario, the "allowed_start_tolerance" was set to be relatively high so it was now upon the JTC to verify that the deviation was within tolerance.

EDIT

This PR also fixes a goal_time tolerance check bug in a scenario when the goal_tolerance was set. It seems that after the last point in a trajectory is commanded, `outside_goal_tolerance` would get set to `true` if the goal_tolerance was violated causing an abort to occur even though `goal_time` was not set! If the goal_time is not set, then by default JTC should wait to complete that the move was successful until the goal_tolerance is reached. This PR fixes this bug by only aborting a violation of the goal tolerance if the `goal_time` has passed and the tolerance has not yet been met as I think was originally intended.

progtologist reviewed (changes requested):
Looks good overall but it would be better if you can add a regression test that verifies the bug and checks so that we don't accidentally repeat the same error in a future commit (regression test).

mechwiz commented:
Ok, I updated the state tolerance test to take into account a more extreme edge case where the desired position is not close to the current position at all and ensured that the position is held after the abort is called.

I also added a test for checking to make sure holding position is working correctly when the goal tolerance is violated.

My changes fix both of these tests. See below

Before my changes:
```
[ RUN      ] TestTrajectoryActions.test_state_tolerances_fail
[INFO] [1648009620.256793835] [test_joint_trajectory_controller]: Command interfaces are [position] and and state interfaces are [position velocity].
[INFO] [1648009620.257763234] [test_joint_trajectory_controller]: Controller state will be published at 50.00 Hz.
[INFO] [1648009620.258758802] [test_joint_trajectory_controller]: Action status changes will be monitored at 20.00 Hz.
[INFO] [1648009620.567656581] [test_joint_trajectory_controller]: Received new action goal
[INFO] [1648009620.567702023] [test_joint_trajectory_controller]: Accepted new action goal
[WARN] [1648009620.567881043] [test_joint_trajectory_controller]: Aborted due to state tolerance violation
/opt/colcon_ws/src/system/ros2_controllers/joint_trajectory_controller/test/test_trajectory_actions.cpp:415: Failure
The difference between init_pos1 and joint_pos_[0] is 0.061260009005657867, which exceeds COMMON_THRESHOLD, where
init_pos1 evaluates to 1.1000000000000001,
joint_pos_[0] evaluates to 1.161260009005658, and
COMMON_THRESHOLD evaluates to 0.0011000000000000001.
/opt/colcon_ws/src/system/ros2_controllers/joint_trajectory_controller/test/test_trajectory_actions.cpp:416: Failure
The difference between init_pos2 and joint_pos_[1] is 0.061260009005657867, which exceeds COMMON_THRESHOLD, where
init_pos2 evaluates to 2.1000000000000001,
joint_pos_[1] evaluates to 2.161260009005658, and
COMMON_THRESHOLD evaluates to 0.0011000000000000001.
/opt/colcon_ws/src/system/ros2_controllers/joint_trajectory_controller/test/test_trajectory_actions.cpp:417: Failure
The difference between init_pos3 and joint_pos_[2] is 0.061260009005657867, which exceeds COMMON_THRESHOLD, where
init_pos3 evaluates to 3.1000000000000001,
joint_pos_[2] evaluates to 3.161260009005658, and
COMMON_THRESHOLD evaluates to 0.0011000000000000001.
[  FAILED  ] TestTrajectoryActions.test_state_tolerances_fail (2434 ms)
[ RUN      ] TestTrajectoryActions.test_goal_tolerances_fail
[INFO] [1648009622.688275054] [test_joint_trajectory_controller]: Command interfaces are [position] and and state interfaces are [position velocity].
[INFO] [1648009622.689102353] [test_joint_trajectory_controller]: Controller state will be published at 50.00 Hz.
[INFO] [1648009622.689958038] [test_joint_trajectory_controller]: Action status changes will be monitored at 20.00 Hz.
[INFO] [1648009622.997295960] [test_joint_trajectory_controller]: Received new action goal
[INFO] [1648009622.997354734] [test_joint_trajectory_controller]: Accepted new action goal
[WARN] [1648009622.997527847] [test_joint_trajectory_controller]: Aborted due to goal tolerance violation
/opt/colcon_ws/src/system/ros2_controllers/joint_trajectory_controller/test/test_trajectory_actions.cpp:462: Failure
The difference between init_pos1 and joint_pos_[0] is 2.8999999999999999, which exceeds COMMON_THRESHOLD, where
init_pos1 evaluates to 1.1000000000000001,
joint_pos_[0] evaluates to 4, and
COMMON_THRESHOLD evaluates to 0.0011000000000000001.
/opt/colcon_ws/src/system/ros2_controllers/joint_trajectory_controller/test/test_trajectory_actions.cpp:463: Failure
The difference between init_pos2 and joint_pos_[1] is 2.8999999999999999, which exceeds COMMON_THRESHOLD, where
init_pos2 evaluates to 2.1000000000000001,
joint_pos_[1] evaluates to 5, and
COMMON_THRESHOLD evaluates to 0.0011000000000000001.
/opt/colcon_ws/src/system/ros2_controllers/joint_trajectory_controller/test/test_trajectory_actions.cpp:464: Failure
The difference between init_pos3 and joint_pos_[2] is 2.8999999999999999, which exceeds COMMON_THRESHOLD, where
init_pos3 evaluates to 3.1000000000000001,
joint_pos_[2] evaluates to 6, and
COMMON_THRESHOLD evaluates to 0.0011000000000000001.
[  FAILED  ] TestTrajectoryActions.test_goal_tolerances_fail (2430 ms)
```

After the changes

```
[ RUN      ] TestTrajectoryActions.test_state_tolerances_fail
[INFO] [1648009578.160415448] [test_joint_trajectory_controller]: Command interfaces are [position] and and state interfaces are [position velocity].
[INFO] [1648009578.161600256] [test_joint_trajectory_controller]: Controller state will be published at 50.00 Hz.
[INFO] [1648009578.162539646] [test_joint_trajectory_controller]: Action status changes will be monitored at 20.00 Hz.
[INFO] [1648009578.470723486] [test_joint_trajectory_controller]: Received new action goal
[INFO] [1648009578.470761977] [test_joint_trajectory_controller]: Accepted new action goal
[WARN] [1648009578.470895820] [test_joint_trajectory_controller]: Aborted due to state tolerance violation
[       OK ] TestTrajectoryActions.test_state_tolerances_fail (2435 ms)
[ RUN      ] TestTrajectoryActions.test_goal_tolerances_fail
[INFO] [1648009580.594484661] [test_joint_trajectory_controller]: Command interfaces are [position] and and state interfaces are [position velocity].
[INFO] [1648009580.595501543] [test_joint_trajectory_controller]: Controller state will be published at 50.00 Hz.
[INFO] [1648009580.596436583] [test_joint_trajectory_controller]: Action status changes will be monitored at 20.00 Hz.
[INFO] [1648009580.904640746] [test_joint_trajectory_controller]: Received new action goal
[INFO] [1648009580.904680487] [test_joint_trajectory_controller]: Accepted new action goal
[WARN] [1648009580.904805627] [test_joint_trajectory_controller]: Aborted due to goal tolerance violation
[       OK ] TestTrajectoryActions.test_goal_tolerances_fail (2434 ms)
```

mechwiz commented:
The only thing I need to figure out is while my changes fix this issue, 4 previous tests now fail. I'll need to investigate why

```
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] PositionVelocityTrajectoryControllers/TrajectoryControllerTestParameterized.correct_initialization_using_parameters/1, where GetParam() = ({ "position", "velocity" }, { "position", "velocity" })
[  FAILED  ] PositionVelocityTrajectoryControllers/TrajectoryControllerTestParameterized.correct_initialization_using_parameters/2, where GetParam() = ({ "position", "velocity" }, { "position", "velocity", "acceleration" })
[  FAILED  ] PositionVelocityAccelerationTrajectoryControllers/TrajectoryControllerTestParameterized.correct_initialization_using_parameters/1, where GetParam() = ({ "position", "velocity", "acceleration" }, { "position", "velocity" })
[  FAILED  ] PositionVelocityAccelerationTrajectoryControllers/TrajectoryControllerTestParameterized.correct_initialization_using_parameters/2, where GetParam() = ({ "position", "velocity", "acceleration" }, { "position", "velocity", "acceleration" })
```

destogl commented:
@mechwiz just closing and reopening to trigger CI

mechwiz commented:
@progtologist , @destogl , What's the protocol for getting PRs merged if CI isn't currently working properly?

destogl commented:
Can you please rebase the PR to current master, the CI will work then.

mechwiz commented:
I have rebased.
Unfortunately, it looks like some of the joint_trajectory_controller tests will fail as @bmagyar described here https://github.com/ros-controls/ros2_controllers/pull/296#issuecomment-1050617186 but on the bright side it looks like no additional tests fail due to my changes from what I can tell.

mechwiz commented on joint_trajectory_controller/src/joint_trajectory_controller.cpp:
```diff
@@ -295,25 +308,18 @@ controller_interface::return_type JointTrajectoryController::update(
-              RCLCPP_WARN(
-                node_->get_logger(), "Aborted due goal_time_tolerance exceeding by %f seconds",
-                difference);
-            }
```
Hmm, In that scenario, no action should be taken since nothing has occurred to cause a failure or a success. So we just continue to the next update

AndyZe replied:
Alright, that makes sense. I'll add a comment to document it.

AndyZe reviewed (approved):
Approving assuming these 2 small suggestions are accepted

mechwiz commented on joint_trajectory_controller/src/joint_trajectory_controller.cpp:
```diff
@@ -295,25 +308,18 @@ controller_interface::return_type JointTrajectoryController::update(
+            RCLCPP_WARN(
+              node_->get_logger(), "Aborted due goal_time_tolerance exceeding by %f seconds",
+              time_difference);
           }
```
added comment in latest commit

mechwiz commented on joint_trajectory_controller/src/joint_trajectory_controller.cpp:
```diff
@@ -241,11 +209,63 @@ controller_interface::return_type JointTrajectoryController::update(
+
+          if (default_tolerances_.goal_time_tolerance != 0.0)
+          {
+            // if we exceed goal_time_toleralance set it to aborted
```
updated this in latest commit

destogl commented:
@mechwiz can you please rebase to master / resolve the conflict caused by merging other PR into master (sorry, we had to start merging somewhere).

Then we are ready, I think.

mechwiz commented:
sure thing! I have rebased :)

mechwiz commented:
@destogl , are we good to merge this in?

destogl reviewed (approved):
This looks good to me! thanks!
